In [1]:
import numpy as np
import pandas as pd
import random

## -----------------------------------------------------PART A-------------------------------------------------------------------

## 1) Keep top 3 frequent values as it is in the series and replace others values by average[Series should have minimum 3 distinct elements].
Example:                                                                                                                                   
Input: [1,3,5,5,6,6,7,7,7]                                                                                                                                 
Output: [5.22,5.22,5,5,6,6,7,7,7]


In [2]:
num_series = pd.Series([1,3,5,5,6,6,7,7,7]) # input list
print("Original Series:") 
print(num_series)
'''
Other than top 3 repeated will be replaced with mean of the list
'''
result = num_series[~num_series.isin(num_series.value_counts().index[:3])] = np.mean(num_series)
print(num_series)

Original Series:
0    1
1    3
2    5
3    5
4    6
5    6
6    7
7    7
8    7
dtype: int64
0    5.222222
1    5.222222
2    5.000000
3    5.000000
4    6.000000
5    6.000000
6    7.000000
7    7.000000
8    7.000000
dtype: float64


## 2) Replace the first character of every element with '\$\' and add “_” at the end of every element present in the list.
Example:                                                                                                                                  
Input: [“abc”,”def”,”ghi”]                                                                                                                                
Output: ["$bc_","$ef_","$hi_"]


In [3]:
list1 = ["abc","def",'ghi']
outlist = [f'${s[1:]}_' for s in list1]
print(outlist)

['$bc_', '$ef_', '$hi_']


## 3) (a)Convert a numpy array to a dataframe of given shape [User given]

In [4]:
a = np.array([[1,2,3],[3,4,5],[4,5,6],[6,7,8],[8,9,10]])
b = pd.DataFrame(a)
b

0  1   2
0  1  2   3
1  3  4   5
2  4  5   6
3  6  7   8
4  8  9  10

##### another way of 3) (a)

In [5]:
n = int(input("Enter Number of Rows: "))
m = int(input("Enter Number of Columns: "))
a = np.array(np.random.rand(n,m)) #rand prints the matrix of given shape with random number
b = pd.DataFrame(a) 
b

Enter Number of Rows: 3
Enter Number of Columns: 5


0         1         2         3         4
0  0.471455  0.477430  0.327092  0.075160  0.870233
1  0.473717  0.995442  0.387904  0.248164  0.256849
2  0.557578  0.859637  0.984758  0.547561  0.066805

## 3 (b)Calculate the number of characters in each word in a series [ User given, series should have at least 5 elements]

In [6]:
a = ['Gummadi', 'Sai', 'Dheeraj', 'Machine Learning', 'Andrej Karpathy']
b = pd.Series(a)
'''
using for loop we go through each and every word in the list and we calculate the length of the word
'''
for i in b:
    count = len(i)
    print(i, count)

Gummadi 7
Sai 3
Dheeraj 7
Machine Learning 16
Andrej Karpathy 15


##### another way of 3(b)

In [7]:
lst = [] # empty list
n = int(input("Enter number of elements : ")) # number of element we want in the list
if n >= 5: # as condition was given minimum 5 elements so n should be min 5
    for i in range(0, n): #we iterate each and every index and place value in every index
        ele = str(input())
        lst.append(ele) #all the elements re appenden to the emptylist
        
b = pd.Series(lst) #converts list to series
#we will got through each an every element in the series and count the len of the word
for i in b:
    count = len(i)
    print("-"*6)
    print("{} has {} charectaristics".format(i, count))

Enter number of elements : 6
Andrej Karpathy
Fei Fei Li
Yan LeCun
Call Newport
Gummadi Sai Dheraj
Xiao Nai
------
Andrej Karpathy has 15 charectaristics
------
Fei Fei Li has 10 charectaristics
------
Yan LeCun has 9 charectaristics
------
Call Newport has 12 charectaristics
------
Gummadi Sai Dheraj has 18 charectaristics
------
Xiao Nai has 8 charectaristics


## 4) Use this dataframe and find out the mean and highest mark for each class

In [8]:
'''
creating data frame using dictionary
class is the class. mark are key's which wil be the column name's
lists are the values

'''
df = pd.DataFrame({'Class': ["A", "A", "B",'B','A','A','B','B'],
                     
                   'Mark': [1,2,3,4,5,6,7,8]})

df

Class  Mark
0     A     1
1     A     2
2     B     3
3     B     4
4     A     5
5     A     6
6     B     7
7     B     8

In [9]:
df.groupby('Class')['Mark'].mean() # mean of each class

Class
A    3.5
B    5.5
Name: Mark, dtype: float64

In [10]:
df.groupby('Class')['Mark'].max() # max of each class

Class
A    6
B    8
Name: Mark, dtype: int64

# ------------------------------------------------PART B----------------------------------------------------

In [11]:
df = pd.read_csv("Titanic.csv")

Columns/Features:                                                                                                                       
1) Passenger id: Unique Id of passenger                                                                                                               
2) Survived: Whether the passenger is survived or not                                                                                                                            
3) Pclass: Ticket Class                                                                                                               
4) Name: Name of Passesnger                                                                                                               
5) Sex: Gender                                                                                                                         
6) Sibsp : Siblings/Spouse board the titanic                                                                                                               
7) Parch: Parents / Children board the titanic                                                                                                               
8) fare: Passenger Fare                                                                                                               
9) cabin: Cabin number                                                                                                               
10) embarked: Port of Embarkation (baorded on ship)

## 1) Find the average fare from each Port of Embarkation

In [12]:
df.groupby('Embarked')['Fare'].mean() # average fare of each Embarkation

Embarked
C    59.954144
Q    13.276030
S    27.079812
Name: Fare, dtype: float64

## 2) Find the survival ratio in each gender in each PClass

In [13]:
df.pivot_table('Survived', index='Sex', columns='Pclass', aggfunc='count')

Pclass    1    2    3
Sex                  
female   94   76  144
male    122  108  347

Survived ratio of (female: male) in Pclass 1 = 94: 122 (males survived more)                                                          
Survived ratio of (female: male) in Pclass 1 = 76: 108                                                                                                              
Survived ratio of (female: male) in Pclass 1 = 144:347

## 3)  Use 3 columns(Name,Age,Ticket) to create a primary key for every row. The column values should be merged using a ‘_’ only and there shouldn’t be any spaces between the characters.

In [14]:
df1 =df['Name'].astype(str)+'_'+ df['Age'].astype(str).astype(str) +'_'+ df['Ticket'].astype(str)

In [15]:
df1

0                 Braund, Mr. Owen Harris_22.0_A/5 21171
1      Cumings, Mrs. John Bradley (Florence Briggs Th...
2           Heikkinen, Miss. Laina_26.0_STON/O2. 3101282
3      Futrelle, Mrs. Jacques Heath (Lily May Peel)_3...
4                   Allen, Mr. William Henry_35.0_373450
                             ...                        
886                    Montvila, Rev. Juozas_27.0_211536
887             Graham, Miss. Margaret Edith_19.0_112053
888    Johnston, Miss. Catherine Helen "Carrie"_nan_W...
889                    Behr, Mr. Karl Howell_26.0_111369
890                      Dooley, Mr. Patrick_32.0_370376
Length: 891, dtype: object

## Find the desired output as shown in the image below
Note : Create the data frame as shown in the image below
Initial:



In [16]:
df = pd.DataFrame({"Animie": ['Shingeki No Kygin', 'Fullmetal Alchemist: Brotherhood'], 
      'Genre':[['Action','Drama','Fantasy','Military', 'Mystery', 'Shounen', 'Super Power'], 
               ['Action','Military', 'Adventure','Comedy','Drama','Magic','Fantasy','Shounen']]})
df

Animie  \
0                 Shingeki No Kygin   
1  Fullmetal Alchemist: Brotherhood   

                                               Genre  
0  [Action, Drama, Fantasy, Military, Mystery, Sh...  
1  [Action, Military, Adventure, Comedy, Drama, M...

## Output 1: (Exception: for loop can be used for this question)

In [17]:
c = df.Genre.apply(lambda x: pd.Series([1] * len(x), index=x)).fillna(0, downcast='infer')

In [18]:
c

Action  Drama  Fantasy  Military  Mystery  Shounen  Super Power  Adventure  \
0       1      1        1         1        1        1            1          0   
1       1      1        1         1        0        1            0          1   

   Comedy  Magic  
0       0      0  
1       1      1

In [19]:
pd.get_dummies(df["Genre"])

TypeError: unhashable type: 'list'

## Output 2: (Avoid using loops)


In [20]:
df1 = df.explode('Genre')

In [21]:
df1 # intermidiate step of output 2

Animie        Genre
0                 Shingeki No Kygin       Action
0                 Shingeki No Kygin        Drama
0                 Shingeki No Kygin      Fantasy
0                 Shingeki No Kygin     Military
0                 Shingeki No Kygin      Mystery
0                 Shingeki No Kygin      Shounen
0                 Shingeki No Kygin  Super Power
1  Fullmetal Alchemist: Brotherhood       Action
1  Fullmetal Alchemist: Brotherhood     Military
1  Fullmetal Alchemist: Brotherhood    Adventure
1  Fullmetal Alchemist: Brotherhood       Comedy
1  Fullmetal Alchemist: Brotherhood        Drama
1  Fullmetal Alchemist: Brotherhood        Magic
1  Fullmetal Alchemist: Brotherhood      Fantasy
1  Fullmetal Alchemist: Brotherhood      Shounen

In [22]:
groups = df1.groupby('Genre')['Animie'].apply(list)

In [23]:
pd.DataFrame(groups) # outpu2

Animie
Genre                                                         
Action       [Shingeki No Kygin, Fullmetal Alchemist: Broth...
Adventure                   [Fullmetal Alchemist: Brotherhood]
Comedy                      [Fullmetal Alchemist: Brotherhood]
Drama        [Shingeki No Kygin, Fullmetal Alchemist: Broth...
Fantasy      [Shingeki No Kygin, Fullmetal Alchemist: Broth...
Magic                       [Fullmetal Alchemist: Brotherhood]
Military     [Shingeki No Kygin, Fullmetal Alchemist: Broth...
Mystery                                    [Shingeki No Kygin]
Shounen      [Shingeki No Kygin, Fullmetal Alchemist: Broth...
Super Power                                [Shingeki No Kygin]